mlflow/pyspark/optuna/study.py

```diff
@@ -0,0 +1,271 @@
+
+            mlflow.set_tracking_uri(mlflow_tracking_env)
+            mlflow_client = MlflowClient(mlflow_tracking_env)
+
+            storage = MlflowStorage(experiment_id=experiment_id)
+            study = optuna.load_study(study_name=study_name, sampler=sampler, storage=storage)
+            num_trials = sum(map(len, iterator))
+
+            try:
+                error_messages = []
+                _optimize_sequential(
+                    study=study,
+                    func=func,
+                    mlflow_client=mlflow_client,
+                    n_trials=num_trials,
+                    timeout=timeout,
+                    catch=catch,
+                    callbacks=callbacks,
+                )
+                error_messages.append(None)
+            except BaseException:
+                traceback_string = traceback.format_exc()
+                error_messages.append(traceback_string)
+            yield pd.DataFrame({"error": error_messages})
+
+        num_tasks = n_trials
+        if n_jobs == -1:
+            n_jobs = num_tasks
+        input_df = self.spark.range(start=0, end=num_tasks, step=1, numPartitions=n_jobs)
+        trial_tag = f"optuna_trial_{study_name}_{experiment_id}"
+        if self._is_spark_connect_mode:
+            self.spark.addTag(trial_tag)
+        else:
+            job_group_id = self.spark.sparkContext.getLocalProperty("spark.jobGroup.id")
+            if job_group_id is None:
+                job_group_id = trial_tag
+                job_group_description = f"optuna_trial_{study_name}"
+                self.spark.sparkContext.setJobGroup(
+                    job_group_id, job_group_description, interruptOnCancel=True
+                )
+        try:
+            input_df.mapInPandas(
+                func=run_task_on_executor_pd,
+                schema="error string",
+            ).collect()
+        except KeyboardInterrupt as e:
+            if self._is_spark_connect_mode:
+                self.spark.interruptTag(trial_tag)
+            else:
+                self.spark.sparkContext.cancelJobGroup(trial_tag)
+            logger.debug("MlflowSparkStudy optimize terminated by user.")
+            self.mlflow_client.set_terminated(self._study_id, "KILLED")
+            raise e
+        if "error" in input_df.columns:
+            error_count = input_df.filter(col("error") != "").count()
+            if error_count > 0:
+                first_non_null_value = input_df.select(first("error", ignorenulls=True)).first()[0]
+                self.mlflow_client.set_terminated(self._study_id, "KILLED")
+                raise ExecutionException(
+                    f"Optimization run for Optuna MlflowSparkStudy failed. "
+                    f"See full error details in the failed MLflow runs. "
+                    f"First trial failure message: {first_non_null_value}"
+                )
+        self.mlflow_client.set_terminated(self._study_id)
```

Comment:
Is `_study_id` equivalent to `run_id`? I think this is confusing. Can we rename `_study_id` to `_run_id`?

Reply:
Got it, it's fine if it causes some issues.